### ISSUE TITLE
Report question LUCAS

### ISSUE THREAD

**Issue by claralb2002** [MLOpsGroup101/StudentPerformanceML#89]
## Working in the cloud

> In the following section we would like to know more about your experience when developing in the cloud.

### Question 17

> **List all the GCP services that you made use of in your project and shortly explain what each service does?**
>
> Recommended answer length: 50-200 words.
>
> Example:
> *We used the following two services: Engine and Bucket. Engine is used for... and Bucket is used for...*
>
> Answer:

- Cloud Logging: We used it to view logs from builds and the running API.
- Cloud Run: We deployed the API container here to serve predictions.
- VM Manager: We used it to manage the VMs that ran some container builds.
- Cloud Build: We used it to build Docker images from our repo.
- Cloud Storage: We stored the dataset and saved model artifacts here.
- Vertex AI: We used it to run training jobs for our model.
- Networking: We relied on it for basic connectivity and firewall rules between services.
- Artifact Registry: We pushed our Docker images here for deployment.
- Container Registry Vulnerability Scanning: It scanned our images for vulnerabilities (we didn’t really use the results).
- Compute Engine: We ran scripts/experiments on VMs when needed.

--- question 17 fill here ---

### Question 18

> **The backbone of GCP is the Compute engine. Explained how you made use of this service and what type of VMs**
> **you used?**
>
> Recommended answer length: 100-200 words.
>
> Example:
> *We used the compute engine to run our ... . We used instances with the following hardware: ... and we started the*
> *using a custom container: ...*
>
> Answer:

We used Compute Engine as a place to run heavier jobs that didn’t fit neatly into serverless services, mainly for building/testing containers. We created VM instances on‑demand, installed our dependencies with uv, and ran the same training and preprocessing scripts as in local dev. The VMs were also helpful when we needed to pull data from Cloud Storage, train, and then push artifacts back for deployment.

For hardware, we used standard CPU VMs (e.g., e2-standard or similar class instances) because our model is lightweight and doesn’t require GPUs (however we also implemented Vertex AI to flex our immense understanding of the course). The VMs were spun up only when needed, then shut down to avoid unnecessary cost. We ran our code either directly on the VM or inside a Docker container built from our project image, depending on the task.

--- question 18 fill here ---

### Question 19

> **Insert 1-2 images of your GCP bucket, such that we can see what data you have stored in it.**
> **You can take inspiration from [this figure](figures/bucket.png).**
>
> Answer:

--- question 19 fill here ---

### Question 20

> **Upload 1-2 images of your GCP artifact registry, such that we can see the different docker images that you have**
> **stored. You can take inspiration from [this figure](figures/registry.png).**
>
> Answer:

--- question 20 fill here ---

### Question 21

> **Upload 1-2 images of your GCP cloud build history, so we can see the history of the images that have been build in**
> **your project. You can take inspiration from [this figure](figures/build.png).**
>
> Answer:

--- question 21 fill here ---

### Question 22

> **Did you manage to train your model in the cloud using either the Engine or Vertex AI? If yes, explain how you did**
> **it. If not, describe why.**
>
> Recommended answer length: 100-200 words.
>
> Example:
> *We managed to train our model in the cloud using the Engine. We did this by ... . The reason we choose the Engine*
> *was because ...*
>
> Answer:

--- question 22 fill here ---

We did manage to train the model in the cloud. We used a manual GitHub Actions trigger that runs Cloud Build, which in turn launches a Vertex AI custom training job with our training container. We intentionally kept it manual (not tied to file changes) so we could control when training jobs ran and avoid accidental cloud costs. Vertex AI handled provisioning and running the container, and the training script saved the model artifacts to our Cloud Storage bucket under models/. We chose this setup because it gave us controlled training (no booming costs) while still letting us run our own Docker-based training code.

### Question 23

> **Did you manage to write an API for your model? If yes, explain how you did it and if you did anything special. If**
> **not, explain how you would do it.**
>
> Recommended answer length: 100-200 words.
>
> Example:
> *We did manage to write an API for our model. We used FastAPI to do this. We did this by ... . We also added ...*
> *to the API to make it more ...*
>
> Answer:

--- question 23 fill here ---

We did manage to write an API for our model using FastAPI. The API loads the trained PyTorch model and the saved preprocessing pipeline on startup, then exposes a /predict endpoint that accepts a list of rows as JSON. Each request is converted into a DataFrame, run through the same preprocessor used during training, and then passed into the model for inference. This keeps inference consistent with training and avoids feature mismatch issues. We also added a simple health check endpoint so we can verify the service is up. For deployment, the API runs inside a Docker container and runs through Cloud Run.

### Question 24

> **Did you manage to deploy your API, either in locally or cloud? If not, describe why. If yes, describe how and**
> **preferably how you invoke your deployed service?**
>
> Recommended answer length: 100-200 words.
>
> Example:
> *For deployment we wrapped our model into application using ... . We first tried locally serving the model, which*
> *worked. Afterwards we deployed it in the cloud, using ... . To invoke the service an user would call*
> *`curl -X POST -F "file=@file.json"<weburl>`*
>
> Answer:

--- question 24 fill here ---

We deployed the API both locally and in the cloud. Locally, we run it with uv run uvicorn src.stuperml.api:app --reload and send requests to http://localhost:8000/predict. For cloud deployment, we containerized the API and deployed it to Cloud Run, which handles scaling and HTTPS endpoints. The container includes the model and preprocessor artifacts, and the service is invoked by sending a POST request with JSON data. 
This setup lets us test quickly on local machines and then use the same build for cloud setup.

### Question 25

> **Did you perform any unit testing and load testing of your API? If yes, explain how you did it and what results for**
> **the load testing did you get. If not, explain how you would do it.**
>
> Recommended answer length: 100-200 words.
>
> Example:
> *For unit testing we used ... and for load testing we used ... . The results of the load testing showed that ...*
> *before the service crashed.*
>
> Answer:

--- question 25 fill here ---

We did unit tests with pytest and FastAPI’s TestClient. We mocked the model so the tests only check that the endpoints respond and the prediction output format is correct. We didn’t run any load testing. If we had time, we would use Locust or k6 to send lots of concurrent requests to /predict and measure response time, error rate, and how many requests the service can handle before slowing down or failing. That would help us decide whether to scale the service or change settings in Cloud Run.

### Question 26

> **Did you manage to implement monitoring of your deployed model? If yes, explain how it works. If not, explain how**
> **monitoring would help the longevity of your application.**
>
> Recommended answer length: 100-200 words.
>
> Example:
> *We did not manage to implement monitoring. We would like to have monitoring implemented such that over time we could*
> *measure ... and ... that would inform us about this ... behaviour of our application.*
>
> Answer:

--- question 26 fill here ---

We did not implement full model monitoring. We can see basic service health through Cloud Logging (errors, request volumes, latency), but we did not add drift detection, performance tracking, or automated alerts tied to model quality. If we had monitoring, we’d track input feature distributions over time, prediction distributions, and a small set of ground‑truth outcomes when available. That would let us detect data drift, degraded accuracy, or unexpected behavior and retrain the model before it impacts users. A big part of why we didn't, is also because we never optimized our model to be any good.
